web3.js/src/transaction.ts

```diff
@@ -17,6 +17,22 @@ import type {CompiledInstruction} from './message';
  */
 export type TransactionSignature = string;
 
+export type TransactionBlockhash = {
+  blockhash?: Blockhash;
+  lastValidBlockHeight?: number;
+};
+
+export type TransactionSignatureBlockhash = {
+  signature: TransactionSignature;
+  blockhash?: Blockhash;
+  lastValidBlockHeight?: number;
+};
```

Comment:
1. You can intersect these two types to make this one (see the `&` operator below)!
2. I'm not sure this belongs in the `Transaction` file. This is more like the shape of the _configuration_ specific to the `confirmTransaction` method. It should live alongside that.

Think of what you're doing as _refactoring_ the `confirmTransaction` method to take a config instead of a signature string. That config can be one of:

* A nonce-based config
* A blockheight-based config
* A signature string (**deprecated**)

If you were to write this from scratch it would look something like this:

```
/** @deprecated Instead, call `confirmTransaction` using a `TransactionConfirmationConfig` */
function confirmTransaction(config: TransactionSignature);
function confirmTransaction(config: BlockheightBasedTransactionConfimationConfig);
function confirmTransaction(config: NonceBasedTransactionConfirmationConfig);
function confirmTransaction(
  config:
    | BlockheightBasedTransactionConfimationConfig
    | NonceBasedTransactionConfirmationConfig
) {
  /* ... */
}
```

Then, you need types that describe the config:

```
type BlockheightBasedTransactionConfimationConfig = {
  signature: TransactionSignature,
} & TransactionBlockhash;
type NonceBasedTransactionConfirmationConfig = {...};
```

Another, maybe better name for these types would be ‘strategy,’ since that's what it _really_ is.

```
function confirmTransaction(strategy: BlockheightBasedConfirmationStrategy | NonceBasedConfirmationStrategy) { ... }
```

Yeah. That makes more sense, to me. What do you think about that?

Reply:
> I'm not sure this belongs in the Transaction file. This is more like the shape of the configuration specific to the confirmTransaction method. It should live alongside that.

I wanted to try to keep things clean, and since we already had `TransactionSignature` defined in `Transaction` it seemed the best place to put it, but I also see your point about keeping it close to the method. 

> Another, maybe better name for these types would be ‘strategy,’ since that's what it really is.

I think this makes sense. I'll move the type definition and implement what you outlined here.